grafana: Fixes to foldersFromFilesStructure option

root-expert opened the pull request:
This PR includes some additional fixes for the `foldersFromFilesStructure` introduced in https://github.com/grafana/grafana-ansible-collection/pull/326 which was incomplete.

1. It correctly handles creating folders from the file structures under `{{ grafana_dashboards_dir }}` while
2. Registers the created and present dashboards folders and adds a task to remove the folders that should not be present if `grafana_provisioning_synced` is enabled
3. Fixes folders permissions to be 0755 and files permissions to be 0644 by using `u=rwX,g=rX,o=rX`

Given a `{{ grafana_dashboards_dir }}` directory structure of:

```
├── dash1.json
├── dash2.json
└── folder
    └── dash3.json
```

the role will create this exact structure under `/var/lib/grafana/dashboards`.

voidquark commented:
I can also see that this could cause issues, as `__dashboards_copied` is being overwritten between two tasks.
In my opinion, this needs to be properly tested for edge cases, since users can provision dashboards both via Grafana.net and from a folder on the deployer machine.

voidquark commented:
@root-expert Have you had a chance to test this? I'd really appreciate it if this could be thoroughly tested before it's merged. I just want to avoid any potential issues that might arise after the merge.

root-expert commented:
@voidquark sorry for the late reply.
Tested this with:
```yaml
---
- hosts: all
  roles:
    - role: grafana.grafana.grafana
      vars:
        grafana_provisioning_synced: true
        grafana_provisioning_dashboards_from_file_structure: true
        grafana_dashboards_dir: "dashboards/"
        grafana_dashboards:
          - dashboard_id: '1860'
            revision_id: '41'
            datasource: 'Prometheus'
        grafana_ini:
          security:
            admin_user: admin
            admin_password: admin
        grafana_datasources:
          - name: "Prometheus"
```

and dashboards folder contains:
```
dashboards
├── dash1.json
└── test
    ├── dash2.json
    ├── dash3.json
    └── subfolder
        └── dash4.json
```

Ansible reports:

```
TASK [grafana.grafana.grafana : Register present and copied folders list] ******
ok: [default] => {"ansible_facts": {"__folders_copied_list": ["/var/lib/grafana/dashboards/test", "/var/lib/grafana/dashboards/test/subfolder"], "__folders_present_list": ["/var/lib/grafana/dashboards/test", "/var/lib/grafana/dashboards/test/subfolder"]}, "changed": false}

TASK [grafana.grafana.grafana : Register present and copied dashboards list] ***
ok: [default] => {"ansible_facts": {"__dashboards_copied_list": ["/var/lib/grafana/dashboards/1860.json", "/var/lib/grafana/dashboards/dash1.json", "/var/lib/grafana/dashboards/test/dash2.json", "/var/lib/grafana/dashboards/test/dash3.json", "/var/lib/grafana/dashboards/test/subfolder/dash4.json"], "__dashboards_present_list": ["/var/lib/grafana/dashboards/1860.json", "/var/lib/grafana/dashboards/dash1.json", "/var/lib/grafana/dashboards/test/dash2.json", "/var/lib/grafana/dashboards/test/dash3.json", "/var/lib/grafana/dashboards/test/subfolder/dash4.json"]}, "changed": false}
```

Removing `dash1.json` ansible reports:
```
TASK [grafana.grafana.grafana : Register present and copied folders list] ******
ok: [default] => {"ansible_facts": {"__folders_copied_list": ["/var/lib/grafana/dashboards/test", "/var/lib/grafana/dashboards/test/subfolder"], "__folders_present_list": ["/var/lib/grafana/dashboards/test", "/var/lib/grafana/dashboards/test/subfolder"]}, "changed": false}

TASK [grafana.grafana.grafana : Register present and copied dashboards list] ***
ok: [default] => {"ansible_facts": {"__dashboards_copied_list": ["/var/lib/grafana/dashboards/1860.json", "/var/lib/grafana/dashboards/test/dash2.json", "/var/lib/grafana/dashboards/test/dash3.json", "/var/lib/grafana/dashboards/test/subfolder/dash4.json"], "__dashboards_present_list": ["/var/lib/grafana/dashboards/1860.json", "/var/lib/grafana/dashboards/dash1.json", "/var/lib/grafana/dashboards/test/dash2.json", "/var/lib/grafana/dashboards/test/dash3.json", "/var/lib/grafana/dashboards/test/subfolder/dash4.json"]}, "changed": false}

TASK [grafana.grafana.grafana : Remove dashboards not present on deployer machine (synchronize)] ***
changed: [default] => (item=/var/lib/grafana/dashboards/dash1.json) => {"ansible_loop_var": "item", "changed": true, "item": "/var/lib/grafana/dashboards/dash1.json", "path": "/var/lib/grafana/dashboards/dash1.json", "state": "absent"}
```

Removing the `subfolder` folder:
```
TASK [grafana.grafana.grafana : Register present and copied folders list] ******
ok: [default] => {"ansible_facts": {"__folders_copied_list": ["/var/lib/grafana/dashboards/test"], "__folders_present_list": ["/var/lib/grafana/dashboards/test", "/var/lib/grafana/dashboards/test/subfolder"]}, "changed": false}

TASK [grafana.grafana.grafana : Register present and copied dashboards list] ***
ok: [default] => {"ansible_facts": {"__dashboards_copied_list": ["/var/lib/grafana/dashboards/1860.json", "/var/lib/grafana/dashboards/test/dash2.json", "/var/lib/grafana/dashboards/test/dash3.json"], "__dashboards_present_list": ["/var/lib/grafana/dashboards/1860.json", "/var/lib/grafana/dashboards/test/dash2.json", "/var/lib/grafana/dashboards/test/dash3.json", "/var/lib/grafana/dashboards/test/subfolder/dash4.json"]}, "changed": false}

TASK [grafana.grafana.grafana : Remove dashboards not present on deployer machine (synchronize)] ***
changed: [default] => (item=/var/lib/grafana/dashboards/test/subfolder/dash4.json) => {"ansible_loop_var": "item", "changed": true, "item": "/var/lib/grafana/dashboards/test/subfolder/dash4.json", "path": "/var/lib/grafana/dashboards/test/subfolder/dash4.json", "state": "absent"}

TASK [grafana.grafana.grafana : Remove folders not present on deployer machine (synchronize)] ***
changed: [default] => (item=/var/lib/grafana/dashboards/test/subfolder) => {"ansible_loop_var": "item", "changed": true, "item": "/var/lib/grafana/dashboards/test/subfolder", "path": "/var/lib/grafana/dashboards/test/subfolder", "state": "absent"}

```

The role seems idempotent apart from `Import grafana.net dashboards`  and `Copy dashboard files` tasks which is to be expected.

Made some change to my PR:

* Removed `json_query` filter as it requires `jmespath` installed. Replaced with vanilla `map` filter
* The playbook handles any trailing `/` in the `grafana_dashboard_dir` variable.

voidquark commented:
Hello, thanks for fixing this. From what I can see, it’s better than what is currently in main.
The only thing that doesn’t seem to work is directory removal.

If you have multiple directories where you store dashboards and deploy them, and then you decide to remove one of these directories locally and re-run the playbook, the old directories remain on the Grafana side but are empty.

Would you be able to fix this part?

root-expert commented:
@voidquark I don't think Grafana supports this currently. Are you aware of this working?
